src/charm.py

```diff
@@ -9,93 +9,377 @@
+        if not self.unit.is_leader:
+            return
+        # update peer databag to trigger the charm status update
+        self._update_active_relations(event, "broken")
+
+    def _set_config_parameters(self, status: StatusBase):
+        """Return if the new parameters must be set."""
+        if isinstance(status, ActiveStatus):
+            return True
+        if isinstance(status, BlockedStatus) and "add relation" in status.message:
+            return True
+        return False
+
+    def _on_config_changed(self, _):
+        """Handle on config changed event."""
+        # Only execute in the unit leader
+        if not self.unit.is_leader():
+            return
+
+        # read new parameters
+        new_database_name = self.model.config.get("database-name", None)
+        new_topic = self.model.config.get("topic-name", None)
+        new_extra_user_roles = self.model.config.get("extra-user-roles", None)
+
+        # get the unit status
+        status = self._get_unit_status()
+
+        # if the status is active set new values otherwise return
+        if not self._set_config_parameters(status):
+            # set status
+            self.unit.status = status
+            return
+
+        # update parameter in the relation databag
+        self.set_secret("app", "database", new_database_name)
+        self.set_secret("app", "topic", new_topic)
+        self.set_secret("app", "extra-user-roles", new_extra_user_roles)
+
+        # update database and extra-user-roles
+        self.mysql.database = self.database or ""
+        self.mysql.extra_user_roles = self.extra_user_roles or ""
+        self.postgresql.database = self.database or ""
+        self.postgresql.extra_user_roles = self.extra_user_roles or ""
+        self.mongodb.database = self.database or ""
+        self.mongodb.extra_user_roles = self.extra_user_roles or ""
+
+        # Update relation databag
+        database_relation_data = {
+            "database": self.database,
+            "extra-user-roles": self.extra_user_roles or "",
+        }
+        self.update_database_relations(database_relation_data)
+
+        # update topic and extra-user-roles
+        self.kafka.topic = self.topic or ""
+        self.kafka.extra_user_roles = self.extra_user_roles or ""
+        # Update relation databag
+        for rel in self.kafka.relations:
+            self.kafka._update_relation_data(
+                rel.id,
+                {
+                    "topic": self.topic,
+                    "extra-user-roles": self.extra_user_roles or "",
+                },
+            )
+
+        # set status
+        self.unit.status = status
+
+    def _get_status_database(self, database_name: str):
+        """Return unit status based on the database parameter read from config."""
+        if database_name:
+            if database_name != self.database:
+                if self.is_database_related:
+                    return BlockedStatus(
```

Comment:
**todo:** Also type-hint return-type here.
**nitpick:** I'm not certain it's clear what this method does from the name+description alone. I would imagine it would return more than just `Blocked` or `None`. Maybe rename to something like `_has_database_changed()`? If I understood it's intent correctly. If so, maybe return a `bool` and have whatever calls it set the status, rather than it being indirectly generated here?



And for @deusebio's comment, maybe something like:

```python
def _get_status_database(self, database_name: str) -> Optional[StatusBase]:
    """Return unit status based on the database parameter read from config."""
    if database name and database_name != self.database and self.is_database_related:
        return BlockedStatus()

    return None